tests/ui/cast.rs

```diff
@@ -61,4 +61,15 @@ fn main() {
     1u32 as usize; // Should not trigger any lint
     1i32 as isize; // Neither should this
     1i32 as usize;
+
+    // Test cast_unnecessary
+    1i32 as i32;
+    1f32 as f32;
+    false as bool;
+    &1i32 as &i32;
+
+    1i32 as i64;      // Should not trigger
```

Comment:
both are cases of inference. `i32::from_str` is different from `FromStr::from_str`, because the latter has some generic inference going on.

Basically the lint should bail if the casted expression had any inference going on. Check whether `expr_ty_adjusted` and `expr_ty` differ.